Take a step toward complete BB domination

staticfloat opened the pull request:
Provide libexpat, libz and p7zip (true 7z for Windows) on all platforms.  This provides a number of benefits:

* Significantly reduces the amount of work in `win-extras`, which I've always thought was an annoying wart in our build system.

* Provides a consistent `7z` executable on all platforms, which can be picked up by `Pkg`/`BinaryProvider` for consistent unzipping/untarring/ungzipping on all platforms.  Note that on Windows, `7z` is a repackaged binary from sourceforge and is the "full" `7z`, whereas on all other platforms it is a `p7zip` build of the standalone `7za` executable with all relevant patches applied.  There are some format capability differences, but for the important things it should be fine.

* We kind of half-heartedly guaranteed to packages like WinRPM (through `LibExpat.jl`) that `libexpat` would always be available.  This is inconsistently true across platforms, so let's fix that by just shipping it everywhere just like any other dependency.

StefanKarpinski commented:
Unless we actually need XML parsing for Pkg, I’d prefer not to include it. Can’t they be a normal BB-provided library?

staticfloat commented:
The XML parsing isn't for Pkg, it's for WinRPM.  I think we could get around this by making a 1.3-only release of WinRPM that uses `Pkg.Artifacts` to download libexpat.  @vtjnash does that seem reasonable to you?

staticfloat committed:
Distribute `p7zip` and `libz` via our BB tarball bundling

This provides a number of benefits:

* Significantly reduces the amount of work in win-extras, which I've
always thought was an annoying wart in our build system.

* Provides a consistent 7z executable on all platforms, which can be
picked up by Pkg/BinaryProvider for consistent
unzipping/untarring/ungzipping on all platforms. Note that on Windows,
7z is a repackaged binary from sourceforge and is the "full" 7z, whereas
on all other platforms it is a p7zip build of the standalone 7za
executable with all relevant patches applied. There are some format
capability differences, but for the important things it should be fine.

staticfloat commented:
With https://github.com/JuliaIO/LibExpat.jl/pull/99 passing tests, I took `libexpat` out of this PR, so it only provides `p7zip` and `libz`.

vtjnash commented on Makefile:
```diff
@@ -585,37 +586,12 @@ test-%: check-whitespace $(JULIA_BUILD_MODE)
-else
-	$(error no win-extras target for ARCH=$(ARCH))
-endif
+	@$(MAKE) -C $(BUILDROOT)/deps install-p7zip
```
this recursive calls to MAKE is race-y, so it's generally a bad idea—but also unnecessary (since the `julia-deps` target already does it)?

staticfloat replied:
`win-extras` does not depend on `julia-deps`, nor do we want it to.  We don't want to e.g. install LLVM when you run `win-extras`, we only want to install `p7zip`; unfortunately we don't have fine-grained dependency control, so the next best thing is recursive make.  `install-p7zip` is a simple enough operation that I think it's not that bad.  `julia-deps` itself is just a (much larger) recursive make call anyway.

vtjnash replied:
eh, ok, it's wrong but not too bad as you say

staticfloat replied:
I think the only way this could be done "right" is if we were to include `deps/Makefile` from `Makefile`, giving us access to all those targets.

vtjnash commented on deps/zlib.mk:
```diff
@@ -0,0 +1,3 @@
+ZLIB_BB_URL_BASE := https://github.com/JuliaBinaryWrappers/Zlib_jll.jl/releases/download/Zlib-v$(ZLIB_VER)+$(ZLIB_BB_REL)
+ZLIB_BB_NAME := Zlib.v$(ZLIB_VER)
+$(eval $(call bb-install,zlib,ZLIB,false))
```
I guess it's unclear to me—are we supposed to support USE_BINARYBUILDER=0 still?

staticfloat replied:
If you're compiling from source, I think it's reasonable to expect that the user will have `libz` installed.  I'm not sure what to do about `7z` though.

staticfloat replied:
To be a little more clear on the dilemma here; building `7z` is easy on all platforms except windows, where we are already forcing users to download binaries.

I think the right middle ground here is to allow the user to compile with `USE_BINARYBUILDER=0`, but note in `README.md` that Julia needs `7z` to use `Pkg`, and if you turn off BinaryBuilder, you're going to need to provide a `7z` on the host machine.   Does that sound reasonable?

staticfloat replied:
@StefanKarpinski pinging you since I think you'll have an opinion here.

StefanKarpinski replied:
As discussed on the Pkg call, yes that sounds reasonable.

staticfloat commented:
Talked this over in the Pkg call; we are going to implement support for `USE_BINARYBUILDER=0` on non-Windows platforms, since that's where people care about it most.

staticfloat commented:
Updated PR: we now provide build recipes for `libz` and `p7zip` for those that wish to disable BB downloading via either `USE_BINARYBUILDER=0` or more selectively via `USE_BINARYBUILDER_ZLIB=0`/`USE_BINARYBUILDER_P7ZIP=0`.  The `p7zip` build is expected to fail on Windows; you have no choice but to use BB or set `USE_SYSTEM_P7ZIP` on Windows.

omus commented:
> Provides a consistent `7z` executable on all platforms

What's the recommended practise for using the `7z` executable? On the current nightly build of Julia (1.4.0-DEV.194) I can't seem to locate the executable.

staticfloat commented:
Try looking in joinpath(Sys.BINDIR, “..”, “libexec”)

On Tue, Sep 24, 2019 at 06:47 Curtis Vogt <notifications@github.com> wrote:

> Provides a consistent 7z executable on all platforms
>
> What's the recommended practise for using the 7z executable? On the
> current nightly build of Julia (1.4.0-DEV.194) I can't seem to locate the
> executable.
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaLang/julia/pull/33125?email_source=notifications&email_token=AAA762BR6XYEVBUQ64LU7PLQLIK7PA5CNFSM4ISRXIH2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7ONFXY#issuecomment-534565599>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAA762DKFYOWHFYZ4NU6YGTQLIK7PANCNFSM4ISRXIHQ>
> .
>

omus commented:
On the Julia nightly for Windows 64-bit the `joinpath(Sys.BINDIR, "..", "libexec")` folder is present but is empty.

ararslan commented:
7z fails to build on FreeBSD with `USE_BINARYBUILDER=0`.

omus commented:
Related: https://github.com/JuliaLang/julia/pull/33592
